packages/datadog-plugin-langchain/src/handlers/chat_model.js

```diff
@@ -0,0 +1,99 @@
+'use strict'
+
+const LangChainHandler = require('./default')
+
+const COMPLETIONS = 'langchain.response.completions'
+
+class LangChainChatModelHandler extends LangChainHandler {
+  getSpanStartTags (ctx, provider) {
+    const tags = {}
+
+    const inputs = ctx.args?.[0]
+
+    for (const messageSetIndex in inputs) {
+      const messageSet = inputs[messageSetIndex]
+
+      for (const messageIndex in messageSet) {
+        const message = messageSet[messageIndex]
+        if (this.isPromptCompletionSampled()) {
+          tags[`langchain.request.messages.${messageSetIndex}.${messageIndex}.content`] =
+            this.truncate(message.content) || ''
+        }
+        tags[`langchain.request.messages.${messageSetIndex}.${messageIndex}.message_type`] = message.constructor.name
+      }
+    }
+
+    const instance = ctx.instance
+    const identifyingParams = (typeof instance._identifyingParams === 'function' && instance._identifyingParams()) || {}
+    for (const [param, val] of Object.entries(identifyingParams)) {
+      if (param.toLowerCase().includes('apikey') || param.toLowerCase().includes('apitoken')) continue
+      if (typeof val === 'object') {
+        for (const [key, value] of Object.entries(val)) {
+          tags[`langchain.request.${provider}.parameters.${param}.${key}`] = value
+        }
+      } else {
+        tags[`langchain.request.${provider}.parameters.${param}`] = val
+      }
+    }
+
+    return tags
+  }
+
+  getSpanEndTags (ctx) {
+    const { result } = ctx
+
+    const tags = {}
+
+    // TODO: do we need token tagging
```

Comment:
We should have token metrics get tagged, I know Langchain should support token metrics for openai at least

Reply:
to clarify this (as my TODO isn't super helpful beyond face value 😅), referring to [this](https://github.com/DataDog/dd-trace-py/blob/48e895244c935c02c88ee14a5eba13eec41ff9cb/ddtrace/contrib/internal/langchain/patch.py#L128-L156) in our LangChain integration, I don't believe we have an equivalent `get_openai_token_cost_for_model` available to us through LangChain. Should we just do the `langchain.tokens.{prompt, completion, total}_tokens` tag on the span here instead and forgo the cost metric?

Reply:
Does langchain expose token metrics on the returned OpenAI response object? If so, should be as simple as extracting those token metrics and setting them on the span tags. Let me know otherwise 👍

Reply:
yep we can grab the token metrics themselves. i'll add them as tags!

Reply:
added this in for `openai` response objects: https://github.com/DataDog/dd-trace-js/pull/4860/commits/0ecb096fa540520bbab41a39b8db53b1b6bf27dd. when I do the `llmobs` integration, i'll try to make it provider-agnostic as we did for the Python LLMObs integration.